packages/flutter_tools/lib/src/ios/devices.dart

```diff
@@ -770,18 +770,22 @@ class IOSDevice extends Device {
       debuggingOptions: debuggingOptions,
     );
 
+    final bool discoverVMUrlFromLogs = vmServiceDiscovery != null && !isWirelesslyConnected;
+
+    // If mDNS fails, don't throw since url may still be findable through vmServiceDiscovery.
     final Future<Uri?> vmUrlFromMDns = MDnsVmServiceDiscovery.instance!.getVMServiceUriForLaunch(
       packageId,
       this,
       usesIpv6: debuggingOptions.ipv6,
       useDeviceIPAsHost: isWirelesslyConnected,
+      throwOnError: !discoverVMUrlFromLogs,
```

Comment:
Instead of piping down a `throwOnError` parameter, what do you think of having `getVMServiceUriForLaunch` throw a `MissingLocalNetworkPermissionsException`? This method could catch that exception, print the instructions, and exit the tool if needed.